File: org.wolfgang/org.wolfgang.contrail-js/src/main/js/Contrail/codec/payload/Decoder.js
/*
 * Copyright (C)2012 D. Plaindoux.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation; either version 2, or (at your option) any
 * later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 */

/*global define*/

define([ "require", "Core/jObj", "IO/jMarshaller" ],
    function (require, jObj, jMarshaller) {
        "use strict";

        function PayloadDecoder() {
            jObj.bless(this, require("Codec/Factory").core.decoder());
            this.buffer = [];
        }

        PayloadDecoder.INT_LEN = 4;

        PayloadDecoder.init = jObj.constructor([], function () {
            return new PayloadDecoder();
        });

        PayloadDecoder.prototype.getNext = jObj.method([], jObj.types.Array, function () {
            var result, payload;

            if (this.buffer.length < PayloadDecoder.INT_LEN) {
                result = [];
            } else {
                payload = jMarshaller.bytesToInt(this.buffer, 0);
                if (this.buffer.length - PayloadDecoder.INT_LEN < payload) {
                    result = [];
                } else {
                    this.buffer.splice(0, PayloadDecoder.INT_LEN);
                    result = this.buffer.splice(0, payload);
                }
            }

            return result;
        });

        PayloadDecoder.prototype.transform = jObj.method([jObj.types.Array], jObj.types.Array, function (bytes) {
            var results = [], result;

            this.buffer = this.buffer.concat(bytes);

            for (result = this.getNext(); result.length > 0; result = this.getNext()) {
                results.push(result);
            }

            return results;
        });

        PayloadDecoder.prototype.finish = jObj.method([], jObj.types.Array, function () {
            if (this.buffer.length > 0) {
                throw jObj.exception("L.array.must.be.empty");
            } else {
                return [];
            }
        });

        return PayloadDecoder.init;
    });

File: org.wolfgang/org.wolfgang.contrail-core/src/main/java/org/wolfgang/contrail/codec/stringify/Decoder.java
/*
 * Copyright (C)2012 D. Plaindoux.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation; either version 2, or (at your option) any
 * later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 */

package org.wolfgang.contrail.codec.stringify;

import java.util.Arrays;
import java.util.List;

import org.wolfgang.contrail.codec.payload.Bytes;
import org.wolfgang.contrail.component.pipeline.transducer.DataTransducer;
import org.wolfgang.contrail.component.pipeline.transducer.DataTransducerException;

/**
 * <code>Decoder</code> is able to transform a string to a Bytes object.
 * 
 * @author Didier Plaindoux
 * @version 1.0
 */
public class Decoder implements DataTransducer<String, Bytes> {

	public Decoder() {
		super();
	}

	@Override
	public List<Bytes> transform(String source) throws DataTransducerException {
		return Arrays.asList(new Bytes(source.getBytes()));
	}

	@Override
	public List<Bytes> finish() throws DataTransducerException {
		return Arrays.asList();
	}
}